### PR TITLE
Make trie structure less memory-hogging

### PR DESCRIPTION
**Pull request by ravenblackx**
Commit Message: Make trie structure less memory-hogging
Additional Description: Previous trie structure uses approximately 8kb per unique character of prefix; in a production environment with about 27000 route prefixes, this amounted to consuming approximately 1.5GB of additional RAM. This version of a trie reduces that memory impact by a factor of over 100, at a cost of a small increase in insert-time and a *very* small increase in lookup-time (still O(1)). Increase in operations are more than paid for by smaller memory footprint = less memory cache destruction, for tables larger than 2000 characters (e.g. 20 entries 100 characters long, 100 entries 20 characters long, etc.)
Trie used for static headers was replaced in #33932 to avoid being subject to the performance impact of this change.
Risk Level: Small. Existing tests still pass, additional tests verify some new edge-case also behaves fine, performance impact is insignificant (a bit worse in cases where it was already very fast, better in cases where it was less fast).
Testing: Added extra unit tests.
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a

Benchmark below was generated with both implementations being tested side-by-side. PR now removes the old 'big' implementation as there is no longer any call for it.

New benchmark (in PR) results, running as `bazel run -c opt test/common/common:utility_trie_speed_test -- --benchmark-repetitions=10 --benchmark_out_format=csv --benchmark_out=somefile.csv`:

Benchmark | BigMean | SmallMean | Difference (>100% means BigTrieEntry was faster)
-- | -- | -- | --
bmTrieLookupsRequestHeaders | 29.7503 | 45.5528 | 153%
bmTrieLookupsResponseHeaders | 21.1678 | 41.4313 | 195%
bmTrieLookups/10/Mixed | 280.286 | 303.578 | 108%
bmTrieLookups/100/Mixed | 419.074 | 351.339 | 83%
bmTrieLookups/1000/Mixed | 1282.69 | 576.878 | 44%
bmTrieLookups/10000/Mixed | 1376.65 | 768.84 | 55%
bmTrieLookups/10/8 | 4.71941 | 12.7108 | 269%
bmTrieLookups/100/8 | 9.93909 | 17.8588 | 179%
bmTrieLookups/1000/8 | 19.2406 | 29.3718 | 152%
bmTrieLookups/10000/8 | 37.3927 | 34.1951 | 91%
bmTrieLookups/10/128 | 511.97 | 1018.77 | 198%
bmTrieLookups/100/128 | 1734.64 | 749.413 | 43%
bmTrieLookups/1000/128 | 2927.7 | 1083.53 | 37%
bmTrieLookups/10000/128 | 4017.93 | 1319.94 | 32%

### CODE REVIEW

**Comment by adisuissa** [comment]
Thanks for starting this!
I know that previous attempts to update this data-structure saw some actual time (not the O notation) increase, which seemed to be high for the header map structure.
IIRC there are some header-map benchmarks. Can you run them as well?

Also adding @jmarantz as he has good suggestions in this area.
/assign @jmarantz @adisuissa

**Review by jmarantz — COMMENTED** [review]
sweet! I will take a look at the code today.

WDYT of adding a perf benchmark?

**Comment by jmarantz** [comment]
What do you think about adding a benchmark?

/wait

**Comment by ravenblackx** [comment]
> What do you think about adding a benchmark?

I added the output of two existing benchmarks (one of which didn't work before or after the change). The first one seems likely to be what you were looking for, and looks like it's ~a wash.

**Comment by adisuissa** [comment]
I think that most (if not all) of the tests in the benchmark do not really exercise the trie.
AFAIK the trie is used to quickly access the O(1) headers. Consider adding new targeted benchmark tests that exercise accessing different paths (for example, short and long keys).

**Comment by ravenblackx** [comment]
Thinking one possibility if it turns out this is a noticeable additional time-cost for headers might be to use distinct TrieEntry templates for headers vs routes, because the memory cost of the headers trie is ~fixed and relatively small and (maybe) highly-trafficked, vs. the memory cost of the routes one is, well, up to 1.5GB in a real-world example and most likely only experiences one lookup per request.

I'll have a go at fixing that other benchmark and maybe adding more and putting together a side-by-side comparison (`--benchmark_format=csv` is a good tip I didn't know about, thanks!)

It'll take a while because this is mostly a side-task when I'm waiting for other things to build/test, and my current main task doesn't have very long pauses.

**Comment by adisuissa** [comment]
> Thinking one possibility if it turns out this is a noticeable additional time-cost for headers might be to use distinct TrieEntry templates for headers vs routes, because the memory cost of the headers trie is ~fixed and relatively small and (maybe) highly-trafficked, vs. the memory cost of the routes one is, well, up to 1.5GB in a real-world example and most likely only experiences one lookup per request.

+1
Changing the header-map has a bigger impact radius, so having 2 trie implementations seems safer.

**Comment by jmarantz** [comment]
I'm not entirely sure we should split out the header-map trie. Can we first see what's suboptimal about your trie for those header-map case and try to fix that? Maybe there's a way to have a single good trie implementation.

Failing that, we could split into 2 alternate trie implementations optmiized for different purposes, and well documented as to which new code should choose, depending on whether they are more like header-map or more like route.

In either case, can we have a new benchmark which shows the improvements you have made?

**Comment by ravenblackx** [comment]
New benchmark (results in the PR description) suggests the new version is faster for everything but lookups of a pretty small number of small keys - I'm guessing because it makes for fewer cache-misses.

**Comment by jmarantz** [comment]
Can you show the benchmark results as a table where there's a column for 'before' and a column for 'after'? Like I said I know it's a bit of toil but not too bad if you use csv.

DId you do anything to make the Iterate benchmark faster in the new version rather than slower, which it was earlier? Maybe we are getting some inconsistent behavior on your machine. You can use `--benchmark_xxx` arguments of varous sorts (try `--help`) to run more iterations and get statistically significant results.

Whenever doing these optimizations I do think it pays to run enough iterations to increase confidence.

**Comment by ravenblackx** [comment]
Benchmarks side-by-sided. Bit weird that `headerMapImplIterate` is the one benchmark with concerning numbers, but as far as I can tell it doesn't even do anything involving a trie during the benchmark's measured code.

**Comment by jmarantz** [comment]
This looks for the routes. I am a little unconvinced we aren't making things a bit worse for header-maps and thus shoiuld have 2 alternate impls with doc explaining when you'd pick which one.

One other minor request: can you add aa percent-improvement column, and with negative percents when things get slower? It might be easier to see at a glance what got slower vs faster.

You might also want to use some more advanced --benchmark options to do a series of interleaved tests and report the std-deviation. It's just different optoins to pass to the test binary.

**Comment by ravenblackx** [comment]
> This looks good for the routes. I am not entirely convinced that we aren't making things a bit worse for header-maps and thus shoiuld have 2 alternate impls with doc explaining when you'd pick which one.

Funny, you were against that solution previously. :)
Done it this way because I tend to agree, though I think if the static header map grew *just a little bit* it would actually be faster to use the smaller trie for that too (and in real-world use that might already be the case because memory-cache-smashing impact in an isolated benchmark may not reflect memory-cache-smashing real-world performance impact.)

> You might also want to use some more advanced `--benchmark_xxx` options to do a series of interleaved tests and report the std-deviation. It's just different options to pass to the test binary.

Made it a mean over 10 executions for the new benchmark; the std-dev was consistently small enough to not really be worth reporting.
The old benchmark is essentially irrelevant now since the change isn't touching header behavior; I added one column to it with the same execution parameters to make sure it was like the "old" column (i.e. that the sugar-only changes were indeed sugar-only), which it is.

**Comment by adisuissa** [comment]
> > This looks good for the routes. I am not entirely convinced that we aren't making things a bit worse for header-maps and thus shoiuld have 2 alternate impls with doc explaining when you'd pick which one.
> 
> Funny, you were against that solution previously. :)

FWIW, it was I who was for introducing this in 2 steps (reducing blast radius).
That said, if there's no impact (or positive impact) for the headers case, then I'd be ok with updating the trie for both.

**Comment by jmarantz** [comment]
My preference is to iterate on the performance for the header-map and use this new impl always. If you think we can get there, I'm all for it.

But I just had a hard time interpreting the data shown as proving it was always at least as fast as it was prviously.

**Comment by ravenblackx** [comment]
> My preference is to iterate on the performance for the header-map and use this new impl always. If you think we can get there, I'm all for it.

Yeah, I don't think there's a viable way to get the performance of FastTrieEntry with the implementation of SmallTrieEntry at the scale of the header static lookup table. (I do think it would be possible to get the header static lookup table to be faster and smaller than a trie, but that would be a separate PR since the point here was more to improve the size, and, it turns out, also performance, of the route-prefix trie.)

> But I just had a hard time interpreting the data shown as proving it was always at least as fast as it was prviously.

As the benchmark now shows, it is not as fast for small datasets. I guess it might be useful to specifically run the benchmark with a dataset that *is the common set of static headers*. Not sure where that's defined.

**Comment by jmarantz** [comment]
OK I will look at the current state. header map performance is important so I have a hard time suggesitng changes that might make someone's critical path slower.

File: source/common/common/utility.h [hunk 1]
```diff
@@ -636,15 +636,100 @@ class WelfordStandardDeviation {
...
+private:
+  std::array<std::unique_ptr<FastTrieEntry>, 256> entries_;
+};
+
+// A SmallTrieEntry aims to be a good balance of performant and
+// space-efficient, by allocating a vector the size of the range of children
+// the node contains. This should be good for most use-cases.
+//
+// For example, a node with children 'a' and 'z' will contain a vector of
+// size 26, containing two values and 24 nulls. A node with only one
```
Comment (jmarantz) [inline source/common/common/utility.h:673]: did you try using a tree rather than a vector? Something like absl::btree_set ?

I'd be interested to see if that might give good performance in all the cases.

File: test/common/common/utility_trie_speed_test.cc [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+  std::vector<size_t> key_selections;
+  for (size_t i = 0; i < 1024; i++) {
+    key_selections.push_back(keyindex_distribution(prng));
+  }
+  size_t key_index = 0;
+  for (auto _ : state) {
+    UNREFERENCED_PARAMETER(_);
+    auto v = trie.find(keys[key_selections[key_index++]]);
+    key_index &= 1023;
```
Comment (jmarantz) [inline test/common/common/utility_trie_speed_test.cc:49]: some comments around this magic would help.

**Review by jmarantz — COMMENTED** [review]
nice! thanks for the benchmark. A few comments now that I've started to look at  code.

File: source/common/common/utility.h [hunk 3]
```diff
@@ -636,15 +636,100 @@ class WelfordStandardDeviation {
   double m2_{0};
 };
 
-template <class Value> struct TrieEntry {
+// FastTrieEntry is used for very simple very fast lookups.
```
[the remark this answers is not in the document]
Comment (ravenblackx) [reply source/common/common/utility.h:639]: I thought about that and couldn't come up with a way that makes sense - is a SparseTrieEntry suitable for sparse data, or itself sparse?
Maybe `BigTrieEntry` and `ScaledTrieEntry`?

File: source/common/common/utility.h (hunk 1 above)
Comment (ravenblackx) [reply source/common/common/utility.h:673]: Lookups through a btree would be multiple steps. O(log(num_keys)) is surely worse than O(1) when the 1 is also very simple.

File: source/common/common/utility.h [hunk 4]
```diff
@@ -636,15 +636,100 @@ class WelfordStandardDeviation {
...
+private:
+  uint8_t min_child_{0};
+  std::vector<std::unique_ptr<SmallTrieEntry>> children_;
 };
 
 /**
  * A trie used for faster lookup with lookup time at most equal to the size of the key.
  */
-template <class Value> struct TrieLookupTable {
+template <template <class> class EntryType, class Value> struct TrieLookupTable {
```
[the remark this answers is not in the document]
Comment (ravenblackx) [reply source/common/common/utility.h:732]: If you do it that way then at the use-sites you have to declare like

`TrieLookupTable<WhateverTrieEntry<std::unique_ptr<X>>, std::unique_ptr<X>>`

versus this way it's

`TrieLookupTable<WhateverTrieEntry, std::unique_ptr<X>>`

**Comment by jmarantz** [comment]
Maybe not if it fits in a cache line?

**Comment by ravenblackx** [comment]
> Maybe not if it fits in a cache line?

The vast majority of vector-based nodes are going to contain vectors of size 1, which (even with the overhead of also containing a start point and the vector's own overhead) would probably still end up smaller than an absl::btree of size 1. I guess I could give it a try after fixing this other stuff.

**Comment by ravenblackx** [comment]
Performance tested a btree_map, it's *similar* to the stretching vector, but slightly slower throughout (and significantly slower at the small end).

File: source/common/common/utility.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (jmarantz) [reply source/common/common/utility.h:732]: What do you think of this idiom:
```
template<class ValueType> class DenseEntry {
 public:
  using Value = ValueType;
};
template<class ValueType> class SparseEntry {
 public:
  using Value = ValueType;
};

template<class EntryType> class MyContainer {
  using Value = typename EntryType::Value;
};

MyContainer<SparseEntry<int>> sparse;
MyContainer<DenseEntry<bool>> dense;
```

I'm fine with whawt you have; just wanted to put this out there as a taste test. I'm not sure why I can't directly use SparseEntry::ValueType from MyContainer without the 'using' nickname.

[the remark this answers is not in the document]
Comment (adisuissa) [reply source/common/common/utility.h:732]: Just to give another alternative, I slightly prefer using explicit aliases instead of letting the user of TrieLookupTable know the possible types of the first parameter.
What I mean is defining the following:
```
template <typename T>
using BigTrieLookupTable = TrieLookupTable<BigTrieEntry, T>;

template <typename T>
using SmallTrieLookupTable = TrieLookupTable<SmallTrieEntry, T>;
```

**Review by adisuissa — COMMENTED** [review]
Thanks!
Just a minor comment.
I'm guessing that runtime guarding it is out of the question, but I just wanted to raise this point as we may be missing how this is going to impact some use-cases.

**Comment by ravenblackx** [comment]
I think it would probably make the most sense to land #33915 first, then #33932, then this one can be modified back to only having one version and one layer of template (and naming it back to TrieEntry/TrieLookupTable) because BigTrieEntry will no longer have any customers.

File: source/common/common/utility.h [hunk 5]
```diff
@@ -636,15 +636,100 @@ class WelfordStandardDeviation {
...
+//
+// By size, having 256 entries for every node makes each node's overhead
+// (excluding values) consume 8KB; even a trie containing only a single
+// prefix "foobar" consumes 56KB.
+// Using ranged vectors like this makes a single prefix "foobar" consume
+// less than 20 bytes per node, for a total of 0.14KB.
+template <class Value> class SmallTrieEntry {
+public:
+  Value value_{};
+  SmallTrieEntry* operator[](uint8_t i) const {
```
[the remark this answers is not in the document]
Comment (jmarantz) [reply source/common/common/utility.h:690]: Maybe have 2 methods: a const method that returns a const ptr, and a non-const method that returns a non-const ptr?

[the remark this answers is not in the document]
Comment (ravenblackx) [reply source/common/common/utility.h:690]: I don't think that even does anything, if both functions have the same behavior? A client that's asking for a const will get the exact same thing from a non-const-returning function as it would if both functions existed.

And it's effectively a private operator, in that the Trie class never exposes a TrieEntry.

If we go with my suggestion of fixing the static header map first then it can be made more explicit that this is private, in that the entry template could be a private member class of the trie template.

[the remark this answers is not in the document]
Comment (adisuissa) [reply source/common/common/utility.h:690]: It's essentially to ensure const-correctness (see https://chromium.googlesource.com/chromium/src/+/main/styleguide/c++/const.md for more details).

[the remark this answers is not in the document]
Comment (ravenblackx) [reply source/common/common/utility.h:690]: Ah, that makes more sense, I thought the suggestion was const and non-const just on the return value, `const X fn() const;` and `X fn();` is much more coherent. Will do.

**Comment by ravenblackx** [comment]
> I think it would probably make the most sense to land #33915 first, then #33932, [...]

Added bonus of this approach, it will fix the msan/tsan problem of OOM-crashing when running the benchmark, because there will no longer be a BigTrieEntry construct to compare, which is the one causing the OOM, which is essentially the thing this PR was intended to fix in the first place. :)

File: source/common/common/utility.h [hunk 6]
```diff
@@ -687,12 +688,15 @@ template <class Value> class BigTrieEntry {
 template <class Value> class SmallTrieEntry {
 public:
   Value value_{};
-  SmallTrieEntry* operator[](uint8_t i) const {
+  const SmallTrieEntry* operator[](uint8_t i) const {
```
Comment (jmarantz) [inline source/common/common/utility.h:691]: If this is never used, what happens if you comment it out?

Comment (ravenblackx) [reply source/common/common/utility.h:691]: Oh yeah, of course it's used because the find() is const, I'm totally wrong.

**Comment by ravenblackx** [comment]
Now that the static header map search isn't a TrieLookupTable, this PR is much simpler because it's back to not needing to support two implementations (and also no longer needs to touch any client-code).

Ready for review again.

File: source/common/common/utility.h [hunk 7]
```diff
@@ -636,9 +636,66 @@ class WelfordStandardDeviation {
...
+
+  void set(uint8_t branch, std::unique_ptr<TrieEntry<Value>> entry) {
+    if (children_.empty()) {
+      children_.reserve(1);
+      children_.emplace_back(std::move(entry));
+      min_child_ = branch;
+      return;
+    }
+    if (branch < min_child_) {
+      // expand the vector backwards, by moving the existing vector onto
```
Comment (adisuissa) [inline source/common/common/utility.h:677]: nit-style: comments start with capital letter, and end with '.'.

**Review by adisuissa — COMMENTED** [review]
Code changes LGTM, thanks!
Left a nit comment about the comments style.

File: source/common/common/utility.h [hunk 8]
```diff
@@ -636,9 +636,66 @@ class WelfordStandardDeviation {
...
+// (excluding values) consume 8KB; even a trie containing only a single
+// prefix "foobar" consumes 56KB.
+// Using ranged vectors like this makes a single prefix "foobar" consume
+// less than 20 bytes per node, for a total of 0.14KB.
+template <class Value> class TrieEntry {
+public:
   Value value_{};
-  std::array<std::unique_ptr<TrieEntry>, 256> entries_;
+  const TrieEntry* operator[](uint8_t i) const {
+    if (i >= min_child_ && i < min_child_ + children_.size()) {
```
Comment (jmarantz) [inline source/common/common/utility.h:662]: Can you add parens to the `x<y+z` pattern, really for the benefit of flawed humans like me who sometimes get paranoid and forgetfull about precedence rules?

File: source/common/common/utility.h [hunk 9]
```diff
@@ -657,10 +714,10 @@ template <class Value> struct TrieLookupTable {
   bool add(absl::string_view key, Value value, bool overwrite_existing = true) {
     TrieEntry<Value>* current = &root_;
     for (uint8_t c : key) {
-      if (!current->entries_[c]) {
-        current->entries_[c] = std::make_unique<TrieEntry<Value>>();
+      if (!(*current)[c]) {
```
Comment (jmarantz) [inline source/common/common/utility.h:717]: nit: can we use `!= nullptr` here? I feel like that might make it clearer to the reader that `current` holds an array of pointers rather than an array of bools.

File: source/common/common/utility.h [hunk 10]
```diff
@@ -636,9 +636,66 @@ class WelfordStandardDeviation {
...
+// Compared to allocating 256 entries for every node, this makes insertions
+// a little bit inefficient (especially insertions in reverse order), but
+// trie lookups remain O(length-of-longest-matching-prefix) with just a
+// couple of very cheap operations extra per step.
+//
+// By size, having 256 entries for every node makes each node's overhead
+// (excluding values) consume 8KB; even a trie containing only a single
+// prefix "foobar" consumes 56KB.
+// Using ranged vectors like this makes a single prefix "foobar" consume
+// less than 20 bytes per node, for a total of 0.14KB.
```
Comment (jmarantz) [inline source/common/common/utility.h:657]: Can you describe in the comment here what this type expects of Value? I see below we have a line `return {}` from a function that returns a Value. I'm assuming what that means is that `Value` must be empty-constructable.

But I see in the previous state of this code in that the function returned nullptr, which makes me think that Value must be a pointer (or unique_ptr or shared_ptr).

**Review by jmarantz — COMMENTED** [review]
flushing comments; and apologies for this taking so long.

Really just some style nits and comment clarifications so far. I'd like to help you land this; I've just been slammed recently.

File: source/common/common/utility.h (hunk 10 above)
Comment (ravenblackx) [reply source/common/common/utility.h:657]: Added in TrieLookupTable comment since TrieEntry isn't even really the public API.

My change to {} rather than nullptr was to reduce the strictness of the constraint - there's no reason it has to be a smart pointer, but it does have to be empty-initializable and comparable. I don't think it really needed a comment saying so, because if you were to try to put any incompatible value in it it won't compile.

File: source/common/common/utility.h (hunk 9 above)
Comment (ravenblackx) [reply source/common/common/utility.h]: Done. (Though this complaint belonged on the *previous* iteration of this code, since using `!` on a pointer wasn't a change, and should also be a linter if we want to insist on this.)

File: source/common/common/utility.h (hunk 8 above)
Comment (ravenblackx) [reply source/common/common/utility.h]: Done.

**Review by adisuissa — DISMISSED** [review]
Thanks, LGTM!
Please fix format.

File: source/common/common/utility.h (hunk 10 above)
Comment (jmarantz) [reply source/common/common/utility.h:657]: What you put there is fine, though I'd prefer more helpful comments than forcing the user to discover what's possible by writing code and then seeing if it compiles :)

But if TrieEntry is no supposed to be accessed by users, can we make it a private nested class of TrieLookupTable?

**Review by jmarantz — APPROVED** [review]
this looks great; just one nit about class naming.

@RyanTheOptimist you want to take a senior pass, as a protocols expert? I realize this isn't a protocols change, but the primary use-case being http headers I thought you might bring helpful context.

File: source/common/common/utility.h [hunk 11]
```diff
@@ -636,13 +636,72 @@ class WelfordStandardDeviation {
...
+template <class Value> class TrieEntry {
+public:
   Value value_{};
-  std::array<std::unique_ptr<TrieEntry>, 256> entries_;
+  const TrieEntry* operator[](uint8_t i) const {
```
Comment (RyanTheOptimist) [inline source/common/common/utility.h:661]: nit: brief comments on these methods, please. (In particular the behavior when "i" is not in the Trie).

File: source/common/common/utility.h [hunk 12]
```diff
@@ -636,13 +636,72 @@ class WelfordStandardDeviation {
...
+    if (i >= min_child_ && i < (min_child_ + children_.size())) {
+      return children_[i - min_child_].get();
+    }
+    return nullptr;
+  }
+  TrieEntry* operator[](uint8_t i) { return const_cast<TrieEntry*>(std::as_const(*this)[i]); }
+
+  void set(uint8_t branch, std::unique_ptr<TrieEntry<Value>> entry) {
```
Comment (RyanTheOptimist) [inline source/common/common/utility.h:669]: For this, please make sure to comment the meaning of `branch`.

**Review by RyanTheOptimist — COMMENTED** [review]
Looks good, modulo a request for some comments and a +1 to @jmarantz's point about the non-public class being moved

File: source/common/common/utility.h (hunk 11 above)
Comment (ravenblackx) [reply source/common/common/utility.h]: Done.

File: source/common/common/utility.h (hunk 12 above)
Comment (ravenblackx) [reply source/common/common/utility.h]: Renamed all the key parameters to key, then rethought that because "key" being used for both an entry key and a child-branch-index made it more confusing if anything, so all the node-child-branch-index arguments are named `c` and commented accordingly.

Also renamed `TrieEntry` to `TrieNode`, because most of them are not actually entries, so now there's distinct concepts instead of blurry confusion.

(Had originally left it as `TrieEntry` because that's how it was, not because I thought it was a good idea, but if we're cleaning up all the existing mess, might as well do more!)

Comment (RyanTheOptimist) [reply source/common/common/utility.h]: Big +1 on the TrieEntry -> TrieNode rename. That seems cleaner.

But I'm less clear about the `c`. I too like the name `key`, but I'm not quite sure I appreciate the distinction between a "entry key and a child-branch-index". Can you say more about this? Is the idea that the "entry key" is the full string in the `TrieLookupTable` whereas `child-branch-index` is the `[0-256)` index of the child in the node? If so, I can see not using key for both. In that case, I think I would suggest `child_index` (since that matches up well with its use and an index into `children_`.) WDYT?

Comment (jmarantz) [reply source/common/common/utility.h]: +1 for child_index.

**Review by RyanTheOptimist — APPROVED** [review]
